### PR TITLE
[stdlib/sys]: install one debug function more times

### PR DESCRIPTION
**Pull request by Pouriya-Jahanbakhsh**
Hi. We can install one function with state `undefined` more times beacuse of [here](https://github.com/erlang/otp/blob/OTP-20.3.2/lib/stdlib/src/sys.erl#L586). Read detailed explanation in [my other PR](https://github.com/erlang/otp/pull/1763).  
This backward compatible PR fixes above bug and supports installing one debug function several times.  

Code snippet:  
```erl
% Get state of 'kernel_sup' process:
1> GetDbgFuncs = fun() -> {_, _, _, [_, _, _, X|_]} = sys:get_status(kernel_sup), X end.
#Fun<erl_eval.20.13309239>

2> GetDbgFuncs().
[]

3> State = undefined.
undefined

%% Always return 'undefined' as state:
4> DbgFunc = fun(_, _, _) -> State end.
#Fun<erl_eval.18.13309239>

%% Old functionality:
5> sys:install(kernel_sup, {DbgFunc, State}).
ok

6> GetDbgFuncs().                            
[{#Fun<erl_eval.18.13309239>,undefined}]

%% Should not be installed again:
7> sys:install(kernel_sup, {DbgFunc, State}).
ok

9> GetDbgFuncs().                            
[{#Fun<erl_eval.18.13309239>,undefined}]

%% New functionality:
10> sys:install(kernel_sup, {id_1, DbgFunc, State}).
ok

%% Cannot be added with id 'id_1' again:
11> sys:install(kernel_sup, {id_1, DbgFunc, State}). 
ok

12> sys:install(kernel_sup, {id_2, DbgFunc, State}).
ok

13> GetDbgFuncs().                                  
[{id_2,{#Fun<erl_eval.18.13309239>,undefined}},
 {id_1,{#Fun<erl_eval.18.13309239>,undefined}},
 {#Fun<erl_eval.18.13309239>,undefined}]

14> sys:remove(kernel_sup, DbgFunc).                
ok

15> GetDbgFuncs().                  
[{id_2,{#Fun<erl_eval.18.13309239>,undefined}},
 {id_1,{#Fun<erl_eval.18.13309239>,undefined}}]

16> sys:remove(kernel_sup, id_2).   
ok

17> GetDbgFuncs().               
[{id_1,{#Fun<erl_eval.18.13309239>,undefined}}]
```

### CODE REVIEW

**Comment by Pouriya-Jahanbakhsh** [comment]
All [tests](https://travis-ci.org/Pouriya-Jahanbakhsh/otp/jobs/366311673) passed for my fork.

**Comment by Pouriya-Jahanbakhsh** [comment]
Functions `sys:install/2-3`  and `sys:remove/2-3` always yield `ok`. Are you agree with yielding `{error, already_exists}` in case of installing existing function and yielding `{error, not_found}` in case of removing unavailable function? If yes, I will change below clauses:  
```erlang
...
debug_cmd({install, {Func, FuncState}}, Debug) ->
    {ok, install_debug(Func, FuncState, Debug)};
debug_cmd({install, {FuncId, Func, FuncState}}, Debug) ->
    {ok, install_debug(FuncId, {Func, FuncState}, Debug)};
debug_cmd({remove, FuncOrFuncId}, Debug) ->
{ok, remove_debug(FuncOrFuncId, Debug)};
...
```
to:
```erlang
...
debug_cmd({install, {Func, FuncState}}, Debug) ->
    case install_debug(Func, FuncState, Debug) of
        Debug -> % Previous Debug =:= Current Debug ->
            {{error, already_exists}, Debug};
        Debug2 ->
            {ok, Debug2}
    end;
debug_cmd({install, {FuncId, Func, FuncState}}, Debug) ->
    case install_debug(FuncId, {Func, FuncState}, Debug) of
        Debug -> % Previous Debug =:= Current Debug ->
            {{error, already_exists}, Debug};
        Debug2 ->
            {ok, Debug2}
    end;
debug_cmd({remove, FuncOrFuncId}, Debug) ->
    case remove_debug(FuncOrFuncId, Debug) of
        Debug -> % Previous Debug =:= Current Debug ->
            {{error, not_found}, Debug};
        Debug2 ->
            {ok, Debug2}
    end;
...
```

File: lib/stdlib/src/sys.erl [hunk 1]
```diff
@@ -582,9 +600,9 @@ trim(N, LogData) ->
 %% Debug structure manipulating functions
 %%-----------------------------------------------------------------
 install_debug(Item, Data, Debug) ->
-    case get_debug2(Item, Debug, undefined) of
-	undefined -> [{Item, Data} | Debug];
-	_ -> Debug
+    case lists:keysearch(Item, 1, Debug) of
```
Comment (sirihansen) [inline lib/stdlib/src/sys.erl]: Sorry for being picky - I'm probably missing something... Why did you change these three lines? Didn't get_debug2/3 do the job?

File: lib/stdlib/src/sys.erl [hunk 2]
```diff
@@ -658,6 +677,8 @@ debug_options([{log_to_file, FileName} | T], Debug) ->
     end;
 debug_options([{install, {Func, FuncState}} | T], Debug) ->
     debug_options(T, install_debug(Func, FuncState, Debug));
+debug_options([{install, {FuncId, Func, FuncState}} | T], Debug) ->
+    debug_options(T, install_debug(FuncId, {Func, FuncState}, Debug));
```
Comment (sirihansen) [inline lib/stdlib/src/sys.erl]: This is just a thought... would it be better to use {FuncId,Func} as key instead of only FuncId? That way you would get some sort if namespacing for your ids. The drawback is that you would need to specify both the id and the fun when removing it again. Did you consider this when implementing? Any thoughts?

File: lib/stdlib/test/sys_SUITE.erl [hunk 3]
```diff
@@ -146,6 +146,20 @@ install(Config) when is_list(Config) ->
...
+    sys:install(?server,{id2, SpyFun, func_state}),    
+    {ok,-1} = (catch public_call(1)),
+    %% We have two SpyFun installed:
+    [{spy_got,{request,1},sys_SUITE_server},
+     {spy_got,{request,1},sys_SUITE_server}] = get_messages(),
+    sys:remove(?server, id1),
+    {ok,-1} = (catch public_call(1)),
+    %% We have one SpyFun installed:
+    [{spy_got,{request,1},sys_SUITE_server}] = get_messages(),
+    sys:no_debug(?server),
```
Comment (sirihansen) [inline lib/stdlib/test/sys_SUITE.erl]: Maybe call public_call/1 after this also? Then next get_messages() should still return the empty list then...

File: lib/stdlib/src/sys.erl [hunk 4]
```diff
@@ -267,33 +268,41 @@ no_debug(Name, Timeout) -> send_system_msg(Name, {debug, no_debug}, Timeout).
 
 -spec install(Name, FuncSpec) -> 'ok' when
       Name :: name(),
-      FuncSpec :: {Func, FuncState},
+      FuncSpec :: {Func, FuncState} | {FuncId, Func, FuncState},
+      FuncId :: _,
```
Comment (sirihansen) [inline lib/stdlib/src/sys.erl]: Please change '_' to term().

File: lib/stdlib/src/sys.erl (hunk 1 above)
Comment (Pouriya-Jahanbakhsh) [reply lib/stdlib/src/sys.erl]: actually `get_debug/2` generates above bug! It yields `undefined` for a function with state `undefined` and we don't want to add it again.

File: lib/stdlib/src/sys.erl (hunk 2 above)
Comment (Pouriya-Jahanbakhsh) [reply lib/stdlib/src/sys.erl]: > The drawback is that you would need to specify both the id and the fun when removing it again.
I prefer sending small things in message passing. If this was in a single process, using `{FuncId, Func}` was good. but we should get these from a client process and we should force client processes to keep `{FuncId, Func}` for removing too.

File: lib/stdlib/src/sys.erl (hunk 1 above)
Comment (sirihansen) [reply lib/stdlib/src/sys.erl]: Ah, yes of course! That's the bugfix. Thanks for clarifying!

File: lib/stdlib/test/sys_SUITE.erl (hunk 3 above)
Comment (Pouriya-Jahanbakhsh) [reply lib/stdlib/test/sys_SUITE.erl]: What line should has another `public_call`? I added it after every installing/removing.

File: lib/stdlib/src/sys.erl (hunk 4 above)
Comment (Pouriya-Jahanbakhsh) [reply lib/stdlib/src/sys.erl]: OMG !

File: lib/stdlib/src/sys.erl (hunk 2 above)
Comment (sirihansen) [reply lib/stdlib/src/sys.erl]: Yes, I think that's ok then.

File: lib/stdlib/test/sys_SUITE.erl (hunk 3 above)
Comment (sirihansen) [reply lib/stdlib/test/sys_SUITE.erl]: I was thinking between `sys:no_debug(?server)` and `[]=get_messages()`, to check that it is turned off again. I'ts a minor though - just thought that the call to `get_messages()` on line 162 would be kind of meaningless otherwise. Or did I miss something more??

Comment (Pouriya-Jahanbakhsh) [reply lib/stdlib/test/sys_SUITE.erl]: Cool. got it.

**Comment by sirihansen** [comment]
There's an old (since forever) bug in the sys_SUITE:install/1 test case, so I just pushed a correction for that.
